ssh/tailssh/tailssh.go

```diff
@@ -1129,18 +1146,22 @@ func (c *conn) matchRule(r *tailcfg.SSHRule) (a *tailcfg.SSHAction, localUser st
 	if c.ruleExpired(r) {
 		return nil, "", nil, errRuleExpired
 	}
+	var userMatchError error
 	if !r.Action.Reject {
 		// For all but Reject rules, SSHUsers is required.
 		// If SSHUsers is nil or empty, mapLocalUser will return an
 		// empty string anyway.
 		localUser = mapLocalUser(r.SSHUsers, c.info.sshUser)
 		if localUser == "" {
-			return nil, "", nil, errUserMatch
+			userMatchError = errUserMatch
 		}
 	}
-	if !c.anyPrincipalMatches(r.Principals) {
+	if userMatchError == nil && !c.anyPrincipalMatches(r.Principals) {
 		return nil, "", nil, errPrincipalMatch
 	}
+	if userMatchError != nil {
+		return nil, "", nil, userMatchError
+	}
```

Comment:
I don't understand this part, why do we need to check `anyPrincipalMatches` if `mapLocalUser` failed?

Reply:
Because if `anyPrincipalMatches` fails, that means the peer doesn't have SSH access to this node at all. The message "tailnet policy does not permit you to SSH as user ..." only makes sense if the tailnet policy does permit you to SSH as a different user.

I can't think of a reason why we need to short-circuit evaluating anyPrincipalMatches, I was just being cautious here. The code would certainly be easier to follow if I just switched the order in which we check principals and user matches.

Reply:
Yeah, I think switching the order would make it much clearer. I can't think of a realistic case where it would return a different result because of this.